Random Search GA

KyleErwin opened the pull request:
Added 
+ Rand search GA object
+ Co responding example 
+ .idea to the gitignore

gpampara commented on ga/src/main/scala/cilib/ga/RandomSearch.scala:
```diff
@@ -0,0 +1,28 @@
+    // The mutation method is applied to all the parents.
+    // The parents are mutated by adding a random value gaussian distribution to
+    // point in their position.
+    def mutation(parents: List[Individual], deviation: Double): RVar[List[Individual]] = {
```
So, mutation can be generalised a little more, right?

I'd prefer us to have a function in the form `A => RVar[A]`, whereby you pass in a Distribution. This way you can pass in the whatever kind of distribution to apply, not just gaussian.

The function that then be applied onto the collection, via a `traverse`

KyleErwin replied:
Sure. I'll get working on that.

gpampara commented on ga/src/main/scala/cilib/ga/RandomSearch.scala:
```diff
@@ -0,0 +1,28 @@
+// Author: Kyle Erwin
+object RandomSearch {
+
+    def ga[S](parentSelection: List[Individual] => RVar[List[Individual]], deviation: Double): List[Individual] => Individual => Step[Double, List[Individual]] =
```
We should rather consider making the this a derivative of the standard GA? you would simply pass in an identity function for any kind of crossover?

KyleErwin replied:
Okay cool. I can do that

KyleErwin commented:
@gpampara I've finished up those edits you asked for. 

for the identity function (used as the crossover) I used RVar.Shuffle. 
` def crossover(selection: List[Individual]): RVar[List[Individual]] = RVar.shuffle(selection)
`
and set `p_c` as 1.0. 

Then I also added the ability to pass a distribution function to the GA to be used by the mutation method. So you can now pass something like 
`val distribution = (position: Double) => Dist.stdNormal.flatMap(x => Dist.gaussian(0, 1.25)).map(_ + position)
`
Meaning the mutation method is now more generic. I can create a new pull request if these changes sound good?

gpampara commented:
@KyleErwin No need to create a new PR :) just commit the changes to your `RandomSearchGA` branch and push to your fork on GitHub. This PR will automatically update with the changes.

You sure about the identity function in the GA is right for the usage you want in the random searching GA? `RVar.point` should be sufficient - we just wanna no nothing, but in the same context.

gpampara commented on ga/src/main/scala/cilib/ga/RandomSearch.scala:
```diff
@@ -0,0 +1,29 @@
+            _position.get(parent).traverse(position => for {
+                newPosition <- distribution(position)
+            } yield newPosition).map(a => _position.set(a)(parent))
+        })
```
Just a note that the lens usage here can be simplified to:

```scala
def mutation(distribution: Double => RVar[Double])(parents: List[Individual[S]]): RVar[List[Individual[S]]] =
  parents.traverse(parent =>
    Lenses._position.modifyF((p: Position[Double]) => p.traverse(distribution))(parent))
```

KyleErwin commented on ga/src/main/scala/cilib/ga/RandomSearch.scala:
```diff
@@ -0,0 +1,29 @@
+import scalaz.syntax.traverse._
+
+// Author: Kyle Erwin
+object RandomSearch {
```
Sure we could do that. I actually think it would be better as we would be sticking to the same theme of how things are defined. /if that makes sense

KyleErwin commented on example/src/main/scala/cilib/example/RandomSearch.scala:
```diff
@@ -20,8 +19,9 @@ object RandomSearchExample extends SafeApp {
     ///////////////////////////////////////////
     val randomSelection = (l: List[Individual]) => RVar.sample(2, l).getOrElse(List.empty[Individual])
-    val ga = RandomSearch.ga(randomSelection, 1.25)
+    val distribution = (position: Double) => Dist.stdNormal.flatMap(x => Dist.gaussian(0, 1.25)).map(_ + position)
```
Ah yes my mistake. Then I could take out the `Dist.stdNormal.flatMap` part and just leave the rest.

KyleErwin commented:
@gpampara I have made a few changes. I removed the random search ga object and implemented it as a method in a defaults file. The example is also up to date with the latest cilib

gpampara commented:
These changes will be pending until #264 gets merged

gpampara commented:
Already on the main branch
